### PR TITLE
Issue 2498: Prevent write access check in prepareCH if it's not needed.

### PR DESCRIPTION
**Pull request by oli0044**
see #2498

### CODE REVIEW

File: core/src/main/java/com/graphhopper/GraphHopper.java [hunk 1]
```diff
@@ -1039,7 +1039,9 @@ else if (prepared.containsKey(profile.getProfile())) {
     }
 
     protected Map<String, PrepareContractionHierarchies.Result> prepareCH(boolean closeEarly, List<CHConfig> configsToPrepare) {
-        ensureWriteAccess();
+        if (!closeEarly && !configsToPrepare.isEmpty()) { // do not check write access if no ch profiles have to be prepared and no ealry closing is set
+            ensureWriteAccess();
+        }
```
Comment (easbar) [inline core/src/main/java/com/graphhopper/GraphHopper.java:1044]: I think we should deal with this situation at an earlier point. When `prepareCH` is called it means that CHs should be prepared (written). So if you meant to do a 'pure load' this should be prevented on a higher level. Either you need to take care of this in your configuration (do not add more CH profiles than there are on disk so no delta imports will run), or `loadOrPrepareCH` should offer something like `deltaImport=true`.

Comment (oli0044) [reply core/src/main/java/com/graphhopper/GraphHopper.java:1044]: Well that is exactely the problem: Although there are no additional CH profiles added, `loadOrPrepareCH` still calls `prepareCH(closeEarly, configsToPrepare)` with an empty `List<CHConfig> configsToPrepare`.

So we could check if `List<CHConfig> configsToPrepare` is empty before calling `prepareCH` altogether... however, inside of `prepareCH` there is some `closeEarly` functionality that flushes and closes the `ghStorage`, which is used when called via `importAndClose() -> process(closeEarly = true) -> postProcessing(closeEarly = true) -> loadOrPrepareCH(closeEarly = true) -> prepareCH(closeEarly = true, configsToPrepare)`.

If I prevent `prepareCH` from being called when there are no ` configsToPrepare`, then the `closeEarly` code is not applied.

Can the `closeEarly` code be ignored? Does it belong in `prepareCH`?

**Comment by michaz** [comment]
I think an incremental improvement (so my customer can load their contraction hierarchies on their read-only filesystem again) is to just remove the `ensureWriteAccess()`.

This now means that in the case where the user said `setAllowWrites(false)` *and* there are CHs to build, this will now be violated.

But if the intention of this option was to let GraphHopper be well-behaved on an actual read-only file system (and it was!), that doesn't matter -- it will just fail when it tries to write, and that is less often than _now_, where it always fails, even if it doesn't actually want to write.

File: core/src/main/java/com/graphhopper/GraphHopper.java (hunk 1 above)
Comment (easbar) [reply core/src/main/java/com/graphhopper/GraphHopper.java:1044]: Ok I see.

> Can the closeEarly code be ignored?  Does it belong in prepareCH?

We should probably not ignore it, but it definitely does not belong in `prepareCH`.

**Comment by easbar** [comment]
Superseded by #2500

**Pull request closed without merging**